apps/fipsinstall.c

```diff
@@ -0,0 +1,430 @@
+
+const OPTIONS fipsinstall_options[] = {
+    {"help", OPT_HELP, '-', "Display this summary"},
+    {OPT_MORE_STR, 0, 0, "e.g: openssl fipsinstall -name fips -section "
+     "fipsinstall -out fips.conf -module ./fips.so -mac HMAC "
+     "-macopt digest:SHA256 -macopt hexkey:00"},
+    {"verify", OPT_VERIFY, '-', "Verification mode, i.e verify a config file "
+     "instead of generating one"},
+    {"in", OPT_IN, 's', "Input config file, used when verifying"},
+    {"out", OPT_OUT, 's', "Output config file, used when generating"},
+    {"module", OPT_MODULE, '>', "File name of the provider module"},
+    {"name", OPT_PROV_NAME, 's', "FIPS provider name"},
+    {"section", OPT_SECTION_NAME, 's',
+     "FIPS Provider config section name (optional)"},
+    {"mac", OPT_MAC, 's', "MAC name"},
+    {"macopt", OPT_MACOPT, 's', "MAC algorithm control parameters in n:v form. "
+                                "See 'Supported Controls' in the EVP_MAC_ docs"},
+    {NULL}
+};
+
+static int mac_ctrl_string(EVP_MAC_CTX *ctx, const char *value)
+{
+    int rv;
+    char *stmp, *vtmp = NULL;
+
+    stmp = OPENSSL_strdup(value);
+    if (stmp == NULL)
+        return -1;
+    vtmp = strchr(stmp, ':');
+    if (vtmp != NULL) {
+        *vtmp = 0;
+        vtmp++;
+    }
+    rv = EVP_MAC_ctrl_str(ctx, stmp, vtmp);
+    OPENSSL_free(stmp);
+    return rv;
+}
+
+static int do_mac(EVP_MAC_CTX *ctx, unsigned char *tmp, BIO *in,
+                  unsigned char *out, size_t *out_len)
+{
+    int ret = 0;
+    int i;
+
+    if (!EVP_MAC_init(ctx))
+        goto err;
+    if (EVP_MAC_size(ctx) > *out_len)
+        goto end;
+    for (;;) {
+        i = BIO_read(in, (char *)tmp, BUFSIZE);
+        if (i < 0)
+            goto err;
+        if (i == 0)
+            break;
+        if (!EVP_MAC_update(ctx, tmp, i))
+            goto err;
+    }
+end:
+    if (!EVP_MAC_final(ctx, out, out_len))
+        goto err;
+    ret = 1;
+err:
+    return ret;
+}
+
+static int load_fips_prov_and_run_self_test(const char *prov_name)
+{
+    int ret = 0;
+    OSSL_PROVIDER *prov = NULL;
+
+    prov = OSSL_PROVIDER_load(NULL, prov_name);
+    if (prov == NULL) {
+        BIO_printf(bio_err, "Failed to load FIPS module\n");
+        goto end;
+    }
+    ret = 1;
+end:
+    OSSL_PROVIDER_unload(prov);
+    return ret;
+}
+
+
+static int print_mac(BIO *bio, const char *label, const unsigned char *mac,
+                     size_t len)
+{
+    int ret;
+    char *hexstr = NULL;
+
+    hexstr = OPENSSL_buf2hexstr(mac, (long)len);
+    if (hexstr == NULL)
+        return 0;
+    ret = BIO_printf(bio, "%s = %s\n", label, hexstr);
+    OPENSSL_free(hexstr);
+    return ret;
+}
+
+static int write_config_header(BIO *out, const char *prov_name,
+                               const char *section)
+{
+    return BIO_printf(out, "openssl_conf = openssl_init\n\n")
+           && BIO_printf(out, "[openssl_init]\n")
+           && BIO_printf(out, "%s = %s\n\n", prov_name, section);
```

Comment:
This is not enough, I'm afraid.  You do need to read `doc/man5/config.pod` (recommended: `perldoc -oman doc/man5/config.pod`).

Essentially, the "header" that you want is this:

```
openssl_conf = openssl_init

[openssl_init]
providers = providers_section

[providers_section]
{prov_name} = {section}
```

It must be this way because the `openssl_init` section (or whatever section name that the starting point `openssl_conf` has as value) is a collection keywords for each module that is being configured.  To configure providers, you have the `providers` keyword with the section for all providers as value, and *that* section holds the collection of providers to be condfigured.

Reply:
yes I found that out by debugging into it :) Took me a while to figure out if it doesnt activate via config that it still caches the provider and its loaded config params..